compiler+runtime/src/cpp/jank/runtime/obj/ratio.cpp

```diff
@@ -0,0 +1,556 @@
+#include <jank/runtime/obj/ratio.hpp>
+#include <fmt/compile.h>
+#include <limits>
+
+namespace jank::runtime
+{
+
+  constexpr native_real EPSILON = std::numeric_limits<native_real>::epsilon();
+
+  obj::ratio_data::static_object(native_integer numerator, native_integer denominator)
+    : numerator{ numerator }
+    , denominator{ denominator }
+  {
+    if(denominator == 0)
+    {
+      throw std::invalid_argument("Ratio denominator cannot be zero.");
+    }
+    simplify();
+  }
+
+  object_ptr obj::ratio::create(native_integer const numerator, native_integer const denominator)
+  {
+    auto ratio_data = make_box<obj::ratio_data>(numerator, denominator);
```

Comment:
`make_box` here is doing a dynamic allocation, via the GC, but `ratio_data` is only used locally here. No need for the GC or dynamic memory. We can just do `obj::ratio_data data{ numerator, denominator };`.

Reply:
we need to call the ctor for ratio_data to simplify the ratio before we can check if the denominator is 1 or not. i changed this to `auto ratio_data = obj::ratio_data(num, denom)`, would this work better?

Reply:
Please change it to the line I specified: `obj::ratio_data data{ numerator, denominator };`

`auto` is meant to help avoid the type when the right hand expression already determines that type. But, in this case, there is no right hand expression. We're just creating a `obj::ratio_data`. Also, we always use direct initialization.

Reply:
sg. updated.